core/src/apps/ethereum/sign_tx.py

```diff
@@ -99,16 +105,94 @@ async def sign_tx(
     return result
 
 
-async def handle_erc20(
+async def confirm_tx_data(
+    msg: MsgInSignTx,
+    defs: Definitions,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+    data_total_len: int,
+) -> None:
+    # function distinguishes between staking / smart contracts / regular transactions
+    from .layout import require_confirm_other_data, require_confirm_tx
+
+    if await handle_staking(msg, defs.network, to_address, maximum_fee, fee_items):
+        return
+
+    # Handle ERC-20, currently only 'transfer' function
+    token, recipient, value = await handle_erc20_transfer(msg, defs, to_address)
+
+    if token is None and data_total_len > 0:
+        await require_confirm_other_data(msg.data_initial_chunk, data_total_len)
+
+    await require_confirm_tx(
+        recipient,
+        value,
+        maximum_fee,
+        fee_items,
+        defs.network,
+        token,
+        bool(msg.chunkify),
+    )
+
+
+async def handle_staking(
+    msg: MsgInSignTx,
+    network: EthereumNetworkInfo,
+    to_address: bytes,
+    maximum_fee: str,
+    fee_items: Iterable[tuple[str, str]],
+) -> bool:
+
+    data_reader = BufferReader(msg.data_initial_chunk)
+    if data_reader.remaining_count() < constants.SC_FUNC_SIG_BYTES:
+        return False
+
+    func_sig = data_reader.read_memoryview(constants.SC_FUNC_SIG_BYTES)
+    if msg.to in constants.ADDRESSES_POOL:
```

Comment:
if the user sends `to` value as non-checksummed address, or omits the 0x prefix, this will fail.

we should probably first decode the address to bytes, use the result here, and specify the pool addresses as bytes too

Reply:
Good catch, thanks. https://github.com/trezor/trezor-firmware/pull/3534/commits/339da2725da19a08e116a30fa035fcb17f31a463